### PR TITLE
Refactor radial basis functions

### PR DESCRIPTION
**Pull request by ArnNag**
## Description
The motivation for this PR is:

1. The SpookyNet radial basis functions do not have a concept of "centers," though the `RadialBasisFunction.compute` abstract method requires implementations to include `centers` as a parameter.
2. The `scale_factor` parameter of  `RadialBasisFunction.compute` is unitful, and inconsistently so in different implementations. For example:
[Here](https://github.com/choderalab/modelforge/blob/9c28f0932b2dbadbf43538bdb24907b2b0a778a4/modelforge/potential/utils.py#L441) it has units nm^-2
[Here](https://github.com/choderalab/modelforge/blob/9c28f0932b2dbadbf43538bdb24907b2b0a778a4/modelforge/potential/utils.py#L452) it has units nm

To address these issues, this PR renames `RadialBasisFunction` to `RadialBasisFunctionCore`, which only takes nondimensionalized distances as input. The nondimensionalization is done in its own abstract method. `scale_factor` is now consistently interpreted as having units nm. 

## Todos
Notable points that this PR has either accomplished or will accomplish.
  - [x] Account for the new interpretation of `scale_factor` when it is initialized.
  - [x] SAKE/PhysNet RBF is difficult to separate into a fully unitless part and will require more thought.
## Questions
- [x] Should we even do this, or is it getting too complicated? An alternative is to simply ignore the centers parameter when implementing `RadialBasisFunction.compute` for SpookyNet.
- [x] What is -15_000 [here](https://github.com/choderalab/modelforge/blob/9c28f0932b2dbadbf43538bdb24907b2b0a778a4/modelforge/potential/utils.py#L557) doing?
  - Answer: seems this is never used 

## Status
- [x] Ready to go

### CODE REVIEW

**Comment by ArnNag** [comment]
After trying to implement the Bernstein polynomials from SpookyNet with this framework, I think it makes more sense to make RadialBasisFunctionCore instantiable with "number_of_radial_basis_functions" as a parameter to RadialBasisFunctionCore.__init__ and RadialBasisFunctionCore as an instance method.

**Review by wiederm — APPROVED** [review]
Thanks @ArnNag , these are great changes and standardize the rbf classes nicely!

File: modelforge/potential/utils.py [hunk 1]
```diff
@@ -476,115 +528,85 @@ def __init__(
...
+            Scalar factor by which to multiply output of radial basis functions.
+        trainable_prefactor: bool, default False
+            Whether prefactor is trainable
+        trainable_centers_and_scale_factors: bool, default False
+            Whether centers and scale factors are trainable.
         """
 
-        super().__init__()
+        super().__init__(GaussianRadialBasisFunctionCore, dtype, prefactor, trainable_prefactor)
         self.number_of_radial_basis_functions = number_of_radial_basis_functions
```
[the remark this answers is not in the document]
Comment (ArnNag) [reply modelforge/potential/utils.py:553]: The intent is for `GaussianRadialBasisFunctionWithScaling` to inherit from the abstract base class `RadialBasisFunction`, with a provided intialization* of `RadialBasisFunctionCore` and implementation of  `RadialBasisFunction.nondimensionalize_distances`, while fixing the signatures for abstract methods for computing means and centers. I don't know what the best practices for multiple inheritance and abstract base classes are—is it acceptable for an abc.ABC subclass to also inherit abc.ABC directly and define its own abstract methods?

File: modelforge/potential/utils.py [hunk 2]
```diff
@@ -424,48 +425,99 @@ def compute_angular_sub_aev(self, vectors12: torch.Tensor) -> torch.Tensor:
...
+        super().__init__()
+        if trainable_prefactor:
+            self.prefactor = nn.Parameter(torch.tensor([prefactor], dtype=dtype))
+        else:
+            self.register_buffer("prefactor", torch.tensor([prefactor], dtype=dtype))
+        self.radial_basis_function = radial_basis_function
 
-        Initializes and contains the logic for computing radial symmetry functions.
+    @abstractmethod
+    def nondimensionalize_distances(self, distances: torch.Tensor) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (ArnNag) [reply modelforge/potential/utils.py:478]: I thought about this since it was annoying that the subtract-and-divide logic was repeated, but I felt it was very confusing for centers and means to have units of nanometers in some subclasses and to be unitless in other subclasses. I wanted to make it more clear that cancellation of units is actually happening via an implied multiplier with the inner exp and everything outside that is unitless. (Yuanqing had actually made this multiplier an explicit parameter called "alpha" in his implementation.)

[the remark this answers is not in the document]
Comment (wiederm) [reply modelforge/potential/utils.py:478]: that makes sense! thanks for clarifying!

File: modelforge/potential/utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (wiederm) [reply modelforge/potential/utils.py:553]: I would have put it in the class signature:
```GaussianRadialBasisFunctionWithScaling(RadialBasisFunction, GaussianRadialBasisFunctionCore)```
but I don't have a strong preference.

[the remark this answers is not in the document]
Comment (ArnNag) [reply modelforge/potential/utils.py:553]: `GaussianRadialBasisFunctionWithScaling` doesn't inherit from `GaussianRadialBasisFunctionCore`, it has an intantiated `GaussianRadialBasisFunctionCore` as an attribute.

**Comment by ArnNag** [comment]
@wiederm 
What is axis 1 (with length 1) in these reference arrays supposed to represent?
https://github.com/choderalab/modelforge/blob/b9fbc2092a0480432550407a78ff6d84882031a9/modelforge/tests/test_schnet.py#L204
https://github.com/choderalab/modelforge/blob/b9fbc2092a0480432550407a78ff6d84882031a9/modelforge/tests/precalculated_values.py#L2228

Is there a reason we want the radial basis functions to be able to take as input arrays with an arbitrary number of axes?

**Comment by ArnNag** [comment]
@wiederm Is it okay to merge this?

**Review by wiederm — APPROVED** [review]

**Comment by wiederm** [comment]
Yes, absolutely! That is a great PR! Thank you!
